### PR TITLE
tests: add external spread backend

### PR DESCRIPTION
**Pull request by fgimenez**
This backend will enable the execution of the spread suite on all-snap running instances, so that the image validation can be automated.

With the current image state, it relies on the previous setup of the default user and network on the instances through console-conf and the execution of `tests/lib/external/prepare-ssh.sh` script included in this changeset. It also needs the `SPREAD_EXTERNAL_ADDRESS` environment variable set on the host with the address and port of the running instance, something like `localhost:8023`.

Currently, the backend includes three systems, `ubuntu-core-16-64`, `ubuntu-core-16-arm-32` and `ubuntu-core-16-arm-64`.

### CODE REVIEW

File: tests/lib/adhoc-ubuntu-core/prepare-instance.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+INSTANCE_IP="${1:-localhost}"
+INSTANCE_PORT="${2:-8022}"
+USER="${3:-$(whoami)}"
+
+execute_remote(){
+    ssh -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -p $INSTANCE_PORT $USER@$INSTANCE_IP "$@"
+}
+
+execute_remote "sudo adduser --extrausers --quiet --disabled-password --gecos '' test"
+execute_remote "(echo test:ubuntu | sudo chpasswd)"
```
Comment (niemeyer) [inline tests/lib/adhoc-ubuntu-core/prepare-instance.sh]: No need for the extra subshell via `(...)`, I believe. Same below.

File: tests/lib/fundamental-snaps.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
Comment (niemeyer) [inline tests/lib/fundamental-snaps.sh]: We might simplify this somewhat by calling this script as `names.sh`, and defining it as:
```
gadget_name=$(snap list | sed -n 's/^\(pc\|pi[23]\|dragonboard\) .*/\1/p')
kernel_name=$gadget_name-kernel

if [ "$kernel_name" = "pi3-kernel" ]; then
        kernel_name=pi2-kernel
fi
```
Then we don't need those assignments on every single place we use this file.

File: spread.yaml [hunk 3]
```diff
@@ -55,6 +55,29 @@ backends:
...
+            ADDRESS $SPREAD_ADHOC_UC_ADDRESS:$SPREAD_ADHOC_UC_PORT
+        discard: |
+            echo "Discarding ad-hoc $SPREAD_SYSTEM"
+        restore: |
+            rm -rf $GOPATH/src
+        systems:
+            - ubuntu-core-16.04-64:
+                username: test
+                password: ubuntu
+            - ubuntu-core-16-device-amd64:
```
Comment (niemeyer) [inline spread.yaml]: What's the point of distinguishing between ubuntu-core-16-64 and ubuntu-core-16-device-amd64? These are both the same system, as far as I can see.

File: tests/main/refresh-devmode-nonproduction/task.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    staginig or production the corresponding remote store is used.
+    When executing against the remote stores the tests rely in the existence of
+    a given snap with an updatable version (version string like 2.0+fake1) in the
+    edge channel.
+
+environment:
+    SNAP_NAME: test-snapd-tools
+    SNAP_VERSION_PATTERN: \d+\.\d+\+fake1
+    BLOB_DIR: $(pwd)/fake-store-blobdir
+    STORE_TYPE: fake
```
Comment (niemeyer) [inline tests/main/refresh-devmode-nonproduction/task.yaml]: This is moving the logic of the task out into `tasks/refresh.sh`, apparently to avoid duplicating it across tasks, but the tasks are all being duplicated to use it. Feels like it's precisely creating the problem it is intended to solve.

I bet we can have one single test file, named simply tests/main/refresh/task.yaml, which contains the real logic inside it, and which has a few variants being executed inside it.

Please ping me if you find any issues in doing so.

File: tests/main/revert-devmode-nonproduction/task.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+summary: Check that revert of a snap in devmode restores devmode
+
+systems: [-ubuntu-core-16-64, -ubuntu-core-16-device-amd64, -ubuntu-core-16-device-arm]
+
+environment:
+    STORE_TYPE: fake
+    BLOB_DIR: $(pwd)/fake-store-blobdir
+
+prepare: |
+    . $TESTSLIB/tasks/revert-devmode.sh
```
Comment (niemeyer) [inline tests/main/revert-devmode-nonproduction/task.yaml]: Same thing.

File: spread.yaml [hunk 6]
```diff
@@ -55,6 +55,29 @@ backends:
                 username: ubuntu
                 password: ubuntu
 
+    adhoc-ubuntu-core:
```
Comment (fgimenez) [inline spread.yaml]: Sure, done

File: spread.yaml [hunk 7]
```diff
@@ -55,6 +55,29 @@ backends:
...
+        type: adhoc
+        environment:
+            SPREAD_ADHOC_UC_ADDRESS: "$(HOST: echo ${SPREAD_ADHOC_UC_ADDRESS:-localhost})"
```
Comment (fgimenez) [inline spread.yaml]: I was reusing them for launching the prepare-instance script, agreed that it's better to merge them for the usage inside spread, done

File: spread.yaml [hunk 8]
```diff
@@ -55,6 +55,29 @@ backends:
...
+            SPREAD_ADHOC_UC_PORT: "$(HOST: echo ${SPREAD_ADHOC_UC_PORT:-8022})"
+        allocate: |
+            echo "Allocating $SPREAD_SYSTEM..."
```
Comment (fgimenez) [inline spread.yaml]: Ok done

File: spread.yaml [hunk 9]
```diff
@@ -55,6 +55,29 @@ backends:
...
+            ADDRESS $SPREAD_ADHOC_UC_ADDRESS:$SPREAD_ADHOC_UC_PORT
+        discard: |
```
Comment (fgimenez) [inline spread.yaml]: Done

File: spread.yaml [hunk 10]
```diff
@@ -55,6 +55,29 @@ backends:
...
+            echo "Discarding ad-hoc $SPREAD_SYSTEM"
+        restore: |
+            rm -rf $GOPATH/src
```
Comment (fgimenez) [inline spread.yaml]: Without this, when i try to run the suite again on the same instance after an execution i get:
```
2016/09/20 10:21:19 Sending project data to external-ubuntu-core:ubuntu-core-16-device-amd64...
2016/09/20 10:21:19 Discarding external-ubuntu-core:ubuntu-core-16-device-amd64, cannot send project data: remote directory /home/gopath/src/github.com/snapcore/snapd is not empty
```
If I try `-keep` I get a password that doesn't work (even if I use the password `ubuntu` declared in the backend it seems to try to connect with the root user), and i can not include `-reuse` without password.

File: spread.yaml (hunk 3 above)
Comment (fgimenez) [inline spread.yaml]: There's a different system mainly because of the `tests/main/ack` test, according to Samuele it is expected to fail on the boards but not on a newly created image, he can explain better. With a different system we can exclude the task from being executed in the boards, please let me know if you have in mind a better way of doing it.

There are other tests, like `tests/main/prepare-image-uboot` or `tests/main/interfaces-network-control` that are excluded from the `ubuntu-core-16-64` system but work fine on the boards, maybe we could enable them in `ubuntu-core-16-64` too, i haven't dug into the reasons about why that was disabled.

Having this additional dummy `ubuntu-core-16.04-64` (note that it is different from `ubuntu-core-16-64`) system here prevents spread from crashing because of tasks without systems defined, when spread will be updated to handle those we can remove it, in the time being it has served to verify that all the rest of changes work as expected and to validate the upcoming images. I've added a `FIXME` notice so that we keep track of it.

File: tests/lib/adhoc-ubuntu-core/prepare-instance.sh (hunk 1 above)
Comment (fgimenez) [inline tests/lib/adhoc-ubuntu-core/prepare-instance.sh]: You are right, the quotes are enough for managing the pipes properly, fixed

File: tests/lib/fundamental-snaps.sh (hunk 2 above)
Comment (fgimenez) [inline tests/lib/fundamental-snaps.sh]: Ok, done

File: tests/main/refresh-devmode-nonproduction/task.yaml (hunk 4 above)
Comment (fgimenez) [inline tests/main/refresh-devmode-nonproduction/task.yaml]: Nope, the main intention of extracting the logic from the tasks is duplicating them, so that we can exclude the non-production tasks from being executed on the boards, they were giving assertion errors. According to @pedronis only the tests that are targeted to production should be executed in the devices.

We had previously variants in place for the different types of stores, if we recover them and move the logic back to the tasks we should somehow skip the non-production tests when executing in the boards, maybe I'm missing something, please let me know WDYT.

**Comment by fgimenez** [comment]
@niemeyer thanks for the review! :) comments addressed, pls take another look when you have the time

File: spread.yaml (hunk 3 above)
Comment (pedronis) [inline spread.yaml]: tests requiring a testing snapd with testing keys don't work against an image that has a normal production snapd with only the production keys

Comment (pedronis) [inline spread.yaml]: device is not a great label for this though,  prod-snapd or something might be better

File: spread.yaml (hunk 10 above)
Comment (niemeyer) [inline spread.yaml]: Yeah, there's a problem indeed. When you provide reuse, spread doesn't know which machine the address belongs to. It finds that out by connecting to it and reading the data it wrote there previously. If it cannot connect, it cannot read the data, and cannot tell how to connect. Oops.

Let me think about that one.

File: tests/main/refresh-devmode-nonproduction/task.yaml (hunk 4 above)
Comment (niemeyer) [inline tests/main/refresh-devmode-nonproduction/task.yaml]: This is related to the prior point about multiple systems which are in fact the same device. The distinction is not in fact a different system, or even a different task. The task is exactly the same, on the same system, and indeed the difference lies in that we want variations of that test. That's what variants are for, as you point out, and I think they remain a better way to implement this test.

You can very easily exclude a particular setup on particular conditions by doing
```
if [ something ]; then exit; fi

Comment (fgimenez) [inline tests/main/refresh-devmode-nonproduction/task.yaml]: @niemeyer ok, in that case we would have a passing test instead of a skipped one, please confirm that's what we want and I'll go for it :)

File: spread.yaml (hunk 10 above)
Comment (niemeyer) [inline spread.yaml]: Spread has been improved to support reuse properly and nicely with this case (and every other, actually). Please let me know how it goes @fgimenez.

File: spread.yaml (hunk 3 above)
Comment (niemeyer) [inline spread.yaml]: As we discussed, it's best to support these variations via a different mechanism instead of different systems. That's the sort of reason why variants have been created, for example.

File: tests/main/refresh-devmode-nonproduction/task.yaml (hunk 4 above)
Comment (niemeyer) [inline tests/main/refresh-devmode-nonproduction/task.yaml]: Yes, I think the small awkwardness of having this as passed is far better than moving actual task logic out of the task and duplicating the test across several different directories.

File: spread.yaml (hunk 10 above)
Comment (fgimenez) [inline spread.yaml]: Totally fine! Thanks :) I've removed that restore step.

File: spread.yaml (hunk 3 above)
Comment (fgimenez) [inline spread.yaml]: Ok, I've updated it to exit on the execute steps in case the store type is fake and the system is ubuntu-core-64*, please take a look

File: tests/lib/adhoc-ubuntu-core/prepare-instance.sh [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
Comment (niemeyer) [inline tests/lib/adhoc-ubuntu-core/prepare-instance.sh]: Can we please move this file to /external/prepare-ssh.sh?

File: tests/lib/prepare.sh [hunk 12]
```diff
@@ -239,9 +239,8 @@ prepare_all_snap() {
     fi
 
     echo "Ensure fundamental snaps are still present"
-    . $TESTSLIB/gadget.sh
-    gadget_name=$(get_gadget_name)
-    for name in $gadget_name ${gadget_name}-kernel ubuntu-core; do
+    . $TESTSLIB/names.sh
+    for name in $gadget_name ${kernel_name} ubuntu-core; do
```
Comment (niemeyer) [inline tests/lib/prepare.sh]: This can be `$kernel_name`, same as `$gadget_name` in the same line.

File: tests/lib/reset.sh [hunk 13]
```diff
@@ -33,10 +33,9 @@ reset_classic() {
 
 reset_all_snap() {
     # remove all leftover snaps
-    . $TESTSLIB/gadget.sh
-    gadget_name=$(get_gadget_name)
+    . $TESTSLIB/names.sh
     for snap in $(ls /snap); do
-        if [ "$snap" = "bin" ] || [ "$snap" = "$gadget_name" ] ||  [ "$snap" = "${gadget_name}-kernel" ] || [ "$snap" = "ubuntu-core" ]; then
+        if [ "$snap" = "bin" ] || [ "$snap" = "$gadget_name" ] ||  [ "$snap" = "${kernel_name}" ] || [ "$snap" = "ubuntu-core" ]; then
```
Comment (niemeyer) [inline tests/lib/reset.sh]: This can be `$kernel_name`, same as `$gadget_name` in the same line.

File: tests/main/refresh-devmode/task.yaml [hunk 14]
```diff
@@ -36,6 +35,9 @@ restore: |
     teardown_store $STORE_TYPE $BLOB_DIR
 
 execute: |
+    # skip the test for nonproduction stores on ubuntu-core systems
+    if [ "$STORE_TYPE" = "fake" -a "$SPREAD_SYSTEM" =~ ubuntu-core-16-* ]; then exit; fi
```
Comment (niemeyer) [inline tests/main/refresh-devmode/task.yaml]: Would read better in three lines rather than one. Can even drop the comment, since it's just re-stating exactly what the code is doing.

File: spread.yaml [hunk 15]
```diff
@@ -57,6 +57,24 @@ backends:
...
 
+    external-ubuntu-core:
+        type: adhoc
+        environment:
+            SPREAD_ADHOC_UC_ADDRESS: "$(HOST: echo ${SPREAD_ADHOC_UC_ADDRESS:-localhost:8022})"
+        allocate: |
+            ADDRESS $SPREAD_ADHOC_UC_ADDRESS
+        systems:
+            # FIXME: remove the ubuntu-core-16.04-64 dummy system once the runner handles tasks without systems assigned
+            - ubuntu-core-16.04-64:
```
Comment (niemeyer) [inline spread.yaml]: What's the comment above about?

This system is a bit awkward, as there's no such thing as an ubuntu-core-16.04. Happy to fix Spread if that's what the comment implies. Not sure about what's broken, though.

File: spread.yaml [hunk 16]
```diff
@@ -57,6 +57,24 @@ backends:
...
+                username: test
+                password: ubuntu
+            - ubuntu-core-16-device-amd64:
```
Comment (niemeyer) [inline spread.yaml]: Per the thread in this PR, I still don't quite see why this is necessary. It would be nice to reuse the ubuntu-core-16-64 system for that instead. Let's please catch up online to discuss it.

File: spread.yaml [hunk 17]
```diff
@@ -57,6 +57,24 @@ backends:
...
+                username: test
+                password: ubuntu
+            - ubuntu-core-16-device-arm:
```
Comment (niemeyer) [inline spread.yaml]: Can we please use the ubuntu-core-16-arm-64 convention here (or *-arm-32, if that's the case).

File: spread.yaml [hunk 18]
```diff
@@ -66,6 +84,33 @@ prepare: |
...
+        cat <<EOF > /etc/systemd/system/snapd.socket.d/local.conf
+    [Unit]
+    StartLimitInterval=0
+    EOF
+        exit 0
+    fi
+
     # apt update is hanging on security.ubuntu.com with IPv6.
     sysctl -w net.ipv6.conf.all.disable_ipv6=1
     trap "sysctl -w net.ipv6.conf.all.disable_ipv6=0" EXIT
```
Comment (niemeyer) [inline spread.yaml]: There were some changes below to make this file compatible with 14.04. Some careful conflict solving will need to be performed here. I suggest looking at the actual diff of that specific change to make sure we don't leave anything out, as we don't yet have 14.04 tests running on Travis.

File: spread.yaml [hunk 19]
```diff
@@ -57,6 +57,24 @@ backends:
                 username: ubuntu
                 password: ubuntu
 
+    external-ubuntu-core:
+        type: adhoc
+        environment:
+            SPREAD_ADHOC_UC_ADDRESS: "$(HOST: echo ${SPREAD_ADHOC_UC_ADDRESS:-localhost:8022})"
```
Comment (niemeyer) [inline spread.yaml]: Can we please name this variable $SPREAD_EXTERNAL_ADDRESS?

File: spread.yaml [hunk 20]
```diff
@@ -66,6 +84,33 @@ prepare: |
...
+    # Disable burst limit so resetting the state quickly doesn't create problems.
+    mkdir -p /etc/systemd/system/snapd.service.d
+    cat <<EOF > /etc/systemd/system/snapd.service.d/local.conf
+    [Unit]
+    StartLimitInterval=0
+    [Service]
+    Environment=SNAPD_DEBUG_HTTP=7 SNAP_REEXEC=0
+    EOF
+
+    # adhoc backend and ubuntu-core system, nothing to setup
```
Comment (niemeyer) [inline spread.yaml]: I suggest dropping this comment. It's not very useful, and actually wrong. We are setting up stuff below.

File: spread.yaml [hunk 21]
```diff
@@ -66,6 +84,33 @@ prepare: |
...
+    if [ "$SPREAD_BACKEND" = external-ubuntu-core ]; then
+        # build test binaries
+        if [ ! -f $GOPATH/bin/snapbuild ]; then
+            mkdir -p $GOPATH/bin
+            snap install --devmode --edge classic
+            classic "sudo apt update && apt install -y git golang-go"
+            classic "GOPATH=$GOPATH go get ..${PROJECT_PATH}/tests/lib/snapbuild"
```
Comment (niemeyer) [inline spread.yaml]: It's great to see this working.

File: spread.yaml [hunk 22]
```diff
@@ -66,6 +84,33 @@ prepare: |
...
+            snap remove classic
+        fi
+        mkdir -p /etc/systemd/system/snapd.socket.d
+        cat <<EOF > /etc/systemd/system/snapd.socket.d/local.conf
```
Comment (niemeyer) [inline spread.yaml]: Why is this needed? And why only here and not on classic?

File: spread.yaml [hunk 23]
```diff
@@ -57,6 +57,24 @@ backends:
                 username: ubuntu
                 password: ubuntu
 
+    external-ubuntu-core:
```
Comment (niemeyer) [inline spread.yaml]: Sorry, I know I've suggested this before, but can we please rename this one further to just `external`? It'll be nicer to see `external:ubuntu-core*` on the output than `external-ubuntu-core:ubuntu-core`.

File: tests/main/refresh/task.yaml [hunk 24]
```diff
@@ -36,6 +35,9 @@ restore: |
     teardown_store $STORE_TYPE $BLOB_DIR
 
 execute: |
+    # skip the test for nonproduction stores on ubuntu-core systems
+    if [ "$STORE_TYPE" = "fake" -a "$SPREAD_SYSTEM" =~ ubuntu-core-16-* ]; then exit; fi
```
Comment (niemeyer) [inline tests/main/refresh/task.yaml]: Same as related comment.

File: tests/main/revert-devmode/task.yaml [hunk 25]
```diff
@@ -25,6 +24,9 @@ restore: |
     teardown_store $STORE_TYPE $BLOB_DIR
 
 execute: |
+    # skip the test for nonproduction stores on ubuntu-core systems
+    if [ "$STORE_TYPE" = "fake" -a "$SPREAD_SYSTEM" =~ ubuntu-core-16-* ]; then exit; fi
```
Comment (niemeyer) [inline tests/main/revert-devmode/task.yaml]: Same as related comment.

File: tests/main/revert/task.yaml [hunk 26]
```diff
@@ -25,6 +24,9 @@ restore: |
     teardown_store $STORE_TYPE $BLOB_DIR
 
 execute: |
+    # skip the test for nonproduction stores on ubuntu-core systems
+    if [ "$STORE_TYPE" = "fake" -a "$SPREAD_SYSTEM" =~ ubuntu-core-16-* ]; then exit; fi
```
Comment (niemeyer) [inline tests/main/revert/task.yaml]: Same as related comment.

File: spread.yaml (hunk 15 above)
Comment (fgimenez) [inline spread.yaml]: Yes, we need changes in spread, currently if you have a backend with specific systems and there's one task that excludes all of them, then you get an error, for instance, if we remove the dummy system above and try to execute the suite you get:

```
fgimenez@arkham:~/Desktop/snappy_validate_image$ spread -v -reuse external:ubuntu-core-16-device-amd64
2016/09/27 09:57:44 Found /home/fgimenez/Desktop/snappy_validate_image/spread.yaml.
error: no systems specified for tests/main/interfaces-locale-control
```
The erroring task is different with each execution, probably related to the random ordering of tasks. Adding the dummy systems allow to execute the suite, `tests/main/interfaces-locale-control` is not executed anyway but doesn't throw any error.

File: spread.yaml (hunk 16 above)
Comment (fgimenez) [inline spread.yaml]: Ok, as Samuele commented previously currently there are tests that should not be executed on ready made all-snap systems, but are ok to be executed on the all-snap image created for `ubuntu-core-16-64`,  for instance `tests/main/ack`.

File: spread.yaml (hunk 17 above)
Comment (fgimenez) [inline spread.yaml]: Ok, I've updated it to `ubuntu-core-16-arm`, there's currently no difference in the tests that are executed for the 32 and 64 bits archs, please let me know if we should duplicate the systems.

File: spread.yaml (hunk 20 above)
Comment (fgimenez) [inline spread.yaml]: Done

File: spread.yaml (hunk 18 above)
Comment (fgimenez) [inline spread.yaml]: Ok, I've double checked that the existing changes are in place and the conflict is resolved.

File: tests/lib/adhoc-ubuntu-core/prepare-instance.sh (hunk 11 above)
Comment (fgimenez) [inline tests/lib/adhoc-ubuntu-core/prepare-instance.sh]: Sure, done

File: tests/lib/prepare.sh (hunk 12 above)
Comment (fgimenez) [inline tests/lib/prepare.sh]: Done

File: tests/lib/reset.sh (hunk 13 above)
Comment (fgimenez) [inline tests/lib/reset.sh]: Done

File: tests/main/refresh-devmode/task.yaml (hunk 14 above)
Comment (fgimenez) [inline tests/main/refresh-devmode/task.yaml]: Done

File: tests/main/refresh/task.yaml (hunk 24 above)
Comment (fgimenez) [inline tests/main/refresh/task.yaml]: Done

File: tests/main/revert-devmode/task.yaml (hunk 25 above)
Comment (fgimenez) [inline tests/main/revert-devmode/task.yaml]: Done

File: tests/main/revert/task.yaml (hunk 26 above)
Comment (fgimenez) [inline tests/main/revert/task.yaml]: Done

File: spread.yaml (hunk 22 above)
Comment (fgimenez) [inline spread.yaml]: Sorry, somehow this response got lost.
On amd64 images I was getting the same errors that we got with the service, now with the socket:
```
+ systemctl start snapd.service snapd.socket
A dependency job for snapd.service failed. See 'journalctl -xe' for details.
Job for snapd.socket failed. See "systemctl status snapd.socket" and "journalctl -xe" for details.
```
and in the journal:
```
Sep 19 16:07:05 localhost.localdomain systemd[1]: snapd.service: Job snapd.service/start failed with result 'dependency'.
Sep 19 16:07:05 localhost.localdomain systemd[1]: snapd.socket: Start request repeated too quickly.
Sep 19 16:07:05 localhost.localdomain systemd[1]: Failed to listen on Socket activation for snappy daemon.
-- Subject: Unit snapd.socket has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit snapd.socket has failed.
```
Setting the burst limit for the socket solves the issue, not sure if there's a better way.

Comment (niemeyer) [inline spread.yaml]: Okay, sounds good, but this doesn't respond the second part of the question: why only here and not on classic?

File: spread.yaml (hunk 17 above)
Comment (niemeyer) [inline spread.yaml]: Yes, let's please duplicate. There's no difference because the actual systems are being externally cooked, so the differences are being encoded elsewhere, but they exist and if we support these systems somewhere else (hopefully we will!) we'll want to run two copies of the tests, one on each of those systems, with different images.

File: spread.yaml (hunk 16 above)
Comment (fgimenez) [inline spread.yaml]: Ok, Samuele is the right person to explain this, for the record this is the error I was getting when naming the system `ubuntu-core-16-64` and executing against a running amd64 instance:
```
2016/09/09 16:46:16 Error executing adhoc-ubuntu-core:ubuntu-core-16-64:tests/main/ack : 
-----
mesg: ttyname failed: Inappropriate ioctl for device
+ echo 'Ack when missing matching key fails'
Ack when missing matching key fails
+ snap ack developer1.account
error: cannot assert: assert failed: no matching public key "XCIC_Wvj9_hiAt0b10sDon74oGr3a6xGODkMZqrj63ZzNYUD5N87-ojjPoeN7f1Y" for signature by "testrootorg"
+ echo 'Ack the test store key'
Ack the test store key
+ snap ack testrootorg-store.account-key
error: cannot assert: assert failed: no matching public key "hIedp1AvrWlcDI4uS_qjoFLzjKl5enu4G2FYJpgB3Pj-tUzGlTQBxMBsBmi-tnJR" for signature by "testrootorg"
```

File: spread.yaml (hunk 22 above)
Comment (fgimenez) [inline spread.yaml]: I guess that it might be because it tries to restart the socket quicker than in other systems, how can we design an experiment to know for sure the root cause?

File: spread.yaml (hunk 15 above)
Comment (fgimenez) [inline spread.yaml]: Thanks a lot, works great

File: spread.yaml (hunk 17 above)
Comment (fgimenez) [inline spread.yaml]: Ok, done

File: spread.yaml (hunk 22 above)
Comment (niemeyer) [inline spread.yaml]: Okay, we're already doing that for snapd.service, so should be okay for the socket.

Let's please do both on the same place, instead of special casing this system and backend.

File: spread.yaml (hunk 16 above)
Comment (niemeyer) [inline spread.yaml]: We discussed this in the standup.

Comment (fgimenez) [inline spread.yaml]: Ok, I've added the changes, instead of using an additional environment variable specific to the unmodified ubuntu-core system (the one in the external backend), which gives an unbounded variable error in the customized system (linode backend) where the variable is not present, I've used `$SPREAD_BACKEND` for determining if a test should be skipped or not. At the moment this just affects two tests, ack and ubuntu-core-device-reg. Please, have a look.

I've also taken the opportunity to review the tests that were disabled for ubuntu-core-16-64 and that  work well on the boards, I've made the required adjustments and now they run in both backends, linode and external.

File: spread.yaml (hunk 22 above)
Comment (fgimenez) [inline spread.yaml]: Sure, done
